### PR TITLE
Add SqlCommandBuilder to unify building RelationalCommands from raw SQL

### PR DESCRIPTION
**Pull request by mikary**
Incremental change for #3110, Start to decouple command building from ISqlStatementExecutor, DRY up user command parameter processing.

### CODE REVIEW

File: test/EntityFramework.Relational.Tests/Storage/SqlCommandBuilderTest.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+            var command = builder.Build("SQL COMMAND TEXT");
+
+            Assert.Equal("SQL COMMAND TEXT", command.CommandText);
+            Assert.Equal(0, command.Parameters.Count);
+        }
+
+        [Fact]
+        public virtual void Builds_RelationalCommand_with_emtpy_parameter_list()
+        {
```
Comment (lajones) [inline test/EntityFramework.Relational.Tests/Storage/SqlCommandBuilderTest.cs]: Nit: typo "emtpy"

File: src/EntityFramework.Relational/Storage/Internal/SqlCommandBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+            _commandBuilderFactory = commandBuilderFactory;
+            _sqlGenerator = sqlGenerator;
+            _parameterNameGeneratorFactory = parameterNameGeneratorFactory;
+        }
+
+        public virtual RelationalCommand Build(
+            [NotNull] string sql,
+            [CanBeNull] IReadOnlyList<object> parameters = null)
+        {
```
Comment (lajones) [inline src/EntityFramework.Relational/Storage/Internal/SqlCommandBuilder.cs]: Nit: Doesn't affect anything but here parameters can be null. But it seems that all the points where you call it you check that it's not null?

**Comment by lajones** [comment]
:shipit:

File: src/EntityFramework.Relational/Storage/Internal/SqlCommandBuilder.cs (hunk 2 above)
Comment (mikary) [inline src/EntityFramework.Relational/Storage/Internal/SqlCommandBuilder.cs]: This is actually intentional, the `SqlCommandBuilder` will be replacing part of the functionality of the `SqlStatementExecutor` which manages both command building and execution. Some existing calls in the `DatabaseCreator`, `HistoryRepository` and a few other classes will make use of the null parameter, but it was impractical to include those changes in this PR without expanding the scope of the PR and including changes that would immediately be reverted / reworked in the next PR.

Comment (lajones) [inline src/EntityFramework.Relational/Storage/Internal/SqlCommandBuilder.cs]: OK. Sounds good.

**Pull request merged**